<a href="https://colab.research.google.com/github/vifirsanova/ML-2026-pt-2/blob/main/beginners/2_trees/trees_hw.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание: cравнение древовидных алгоритмов на NLP-данных

**Задача:** oбучить разные древовидные модели на текстах новостей, замерить test accuracy и время обучения, заполнить таблицы.

## 1. Загрузка и подготовка данных (код дан)

In [ ]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import time
import pandas as pd

# Загружаем 3 категории
categories = ['rec.sport.baseball', 'sci.space', 'comp.graphics']
newsgroups = fetch_20newsgroups(subset='all', categories=categories, shuffle=True, random_state=42)

# TF-IDF векторизация
vectorizer = TfidfVectorizer(max_features=5000, ngram_range=(1, 2), stop_words='english')
X = vectorizer.fit_transform(newsgroups.data)
y = newsgroups.target

# Разделение на train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Размер обучающей выборки: {X_train.shape}")
print(f"Размер тестовой выборки: {X_test.shape}")

Размер обучающей выборки: (2363, 5000)
Размер тестовой выборки: (591, 5000)


## 2. Образец: Дерево решений (Decision Tree)

**Гиперпараметр:** `max_depth`

In [ ]:
from sklearn.tree import DecisionTreeClassifier

results_tree = []

# Образец: мы делаем перебор значений max_depth = [3, 5, 10, 20, None])
# Для каждого: замер времени, обучение, accuracy на тесте, сохранение в results_tree

for depth in [3, 5, 10, 20, None]:
    start = time.time()
    clf = DecisionTreeClassifier(max_depth=depth, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_tree.append({'max_depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_tree = pd.DataFrame(results_tree)
print("\nТаблица 1. Дерево решений")
print(df_tree.to_markdown(index=False))

depth=3: acc=0.6244, time=0.22s
depth=5: acc=0.6971, time=0.38s
depth=10: acc=0.7716, time=0.66s
depth=20: acc=0.8409, time=0.98s
depth=None: acc=0.8697, time=1.08s

Таблица 1. Дерево решений
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           3 |          0.6244 |       0.22 |
|           5 |          0.6971 |       0.38 |
|          10 |          0.7716 |       0.66 |
|          20 |          0.8409 |       0.98 |
|         nan |          0.8697 |       1.08 |


## 3. Заполните таблицы для остальных алгоритмов

По аналогии с образцом обучите следующие модели и запишите результаты.

### 3.1. Случайный лес (Random Forest)

**Гиперпараметр:** `max_depth` (фиксируем `n_estimators=100`)

| max_depth | test_accuracy | time_sec |
|-----------|---------------|----------|
| 5         |               |          |
| 10        |               |          |
| 20        |               |          |
| None      |               |          |

In [ ]:
from sklearn.ensemble import RandomForestClassifier

results_rf = []

# Ваш код здесь (перебор max_depth)
for depth in [5, 10, 20, None]:
    start = time.time()
    clf = RandomForestClassifier(max_depth=depth, n_estimators=100, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_rf.append({'max_depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_rf = pd.DataFrame(results_rf)
print("\nТаблица 2. Случайный лес")
print(df_rf.to_markdown(index=False))

depth=5: acc=0.9002, time=0.38s
depth=10: acc=0.9154, time=0.63s
depth=20: acc=0.9357, time=1.36s
depth=None: acc=0.9526, time=2.64s

Таблица 2. Случайный лес
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           5 |          0.9002 |       0.38 |
|          10 |          0.9154 |       0.63 |
|          20 |          0.9357 |       1.36 |
|         nan |          0.9526 |       2.64 |


### 3.2. Градиентный бустинг (Gradient Boosting)

**Гиперпараметр:** `learning_rate` (фиксируем `n_estimators=100, max_depth=3`)

| learning_rate | test_accuracy | time_sec |
|---------------|---------------|----------|
| 0.01          |               |          |
| 0.05          |               |          |
| 0.1           |               |          |
| 0.5           |               |          |

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

results_gb = []

# Ваш код здесь (перебор learning_rate)
for learning_rate in [0.01, 0.05, 0.1, 0.5]:
    start = time.time()
    clf = GradientBoostingClassifier(learning_rate=learning_rate, max_depth=3, n_estimators=100, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_gb.append({'learning_rate': learning_rate, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"learning_rate={learning_rate}: acc={acc:.4f}, time={fit_time:.2f}s")

df_gb = pd.DataFrame(results_gb)
print("\nТаблица 3. Градиентный бустинг")
print(df_gb.to_markdown(index=False))

learning_rate=0.01: acc=0.8782, time=37.65s
learning_rate=0.05: acc=0.9475, time=44.21s
learning_rate=0.1: acc=0.9577, time=38.42s
learning_rate=0.5: acc=0.9662, time=34.65s

Таблица 3. Градиентный бустинг
|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|            0.01 |          0.8782 |      37.65 |
|            0.05 |          0.9475 |      44.21 |
|            0.1  |          0.9577 |      38.42 |
|            0.5  |          0.9662 |      34.65 |


### 3.3. XGBoost

**Гиперпараметр:** `gamma` (фиксируем `n_estimators=100, learning_rate=0.1`)

| gamma | test_accuracy | time_sec |
|-------|---------------|----------|
| 0     |               |          |
| 0.1   |               |          |
| 0.5   |               |          |
| 1.0   |               |          |

In [ ]:
from xgboost import XGBClassifier

results_xgb = []

# Ваш код здесь (перебор gamma, use_label_encoder=False, eval_metric='logloss')
for gamma in [0, 0.1, 0.5, 1.0]:
    start = time.time()
    clf = XGBClassifier(gamma=gamma, n_estimators=100, learning_rate=0.1, random_state=42, eval_metric='logloss')
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_xgb.append({'gamma': gamma, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"gamma={gamma}: acc={acc:.4f}, time={fit_time:.2f}s")

df_xgb = pd.DataFrame(results_xgb)
print("\nТаблица 4. XGBoost")
print(df_xgb.to_markdown(index=False))

gamma=0: acc=0.9459, time=70.33s
gamma=0.1: acc=0.9442, time=49.18s
gamma=0.5: acc=0.9492, time=31.45s
gamma=1.0: acc=0.9492, time=32.54s

Таблица 4. XGBoost
|   gamma |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|     0   |          0.9459 |      70.33 |
|     0.1 |          0.9442 |      49.18 |
|     0.5 |          0.9492 |      31.45 |
|     1   |          0.9492 |      32.54 |


### 3.4. AdaBoost

**Гиперпараметр:** `learning_rate` (фиксируем `n_estimators=100`)

| learning_rate | test_accuracy | time_sec |
|---------------|---------------|----------|
| 0.5           |               |          |
| 1.0           |               |          |
| 1.5           |               |          |
| 2.0           |               |          |

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

results_ada = []

# Ваш код здесь (перебор learning_rate)
for learning_rate in [0.5, 1.0, 1.5, 2.0]:
    start = time.time()
    clf = AdaBoostClassifier(learning_rate=learning_rate, n_estimators=100, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_ada.append({'learning_rate': learning_rate, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"learning_rate={learning_rate}: acc={acc:.4f}, time={fit_time:.2f}s")

df_ada = pd.DataFrame(results_ada)
print("\nТаблица 5. AdaBoost")
print(df_ada.to_markdown(index=False))

learning_rate=0.5: acc=0.8629, time=10.00s
learning_rate=1.0: acc=0.9205, time=7.05s
learning_rate=1.5: acc=0.9255, time=6.12s
learning_rate=2.0: acc=0.8731, time=5.26s

Таблица 5. AdaBoost
|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|             0.5 |          0.8629 |      10    |
|             1   |          0.9205 |       7.05 |
|             1.5 |          0.9255 |       6.12 |
|             2   |          0.8731 |       5.26 |


### 3.5. LightGBM

**Гиперпараметр:** `num_leaves` (фиксируем `n_estimators=100, learning_rate=0.1`)

| num_leaves | test_accuracy | time_sec |
|------------|---------------|----------|
| 15         |               |          |
| 31         |               |          |
| 63         |               |          |
| 127        |               |          |

In [ ]:
from lightgbm import LGBMClassifier

results_lgbm = []

# Ваш код здесь (перебор num_leaves, n_jobs=-1)
for num_leaves in [15, 31, 63, 127]:
    start = time.time()
    clf = LGBMClassifier(num_leaves=num_leaves, n_estimators=100, learning_rate=0.1, verbosity=-1, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_lgbm.append({'num_leaves': num_leaves, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"num_leaves={num_leaves}: acc={acc:.4f}, time={fit_time:.2f}s")

df_lgbm = pd.DataFrame(results_lgbm)
print("\nТаблица 6. LightGBM")
print(df_lgbm.to_markdown(index=False))

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=15: acc=0.9662, time=4.96s


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=31: acc=0.9594, time=7.72s


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=63: acc=0.9577, time=11.29s
num_leaves=127: acc=0.9594, time=15.74s

Таблица 6. LightGBM
|   num_leaves |   test_accuracy |   time_sec |
|-------------:|----------------:|-----------:|
|           15 |          0.9662 |       4.96 |
|           31 |          0.9594 |       7.72 |
|           63 |          0.9577 |      11.29 |
|          127 |          0.9594 |      15.74 |


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


In [14]:
pip install catboost

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.1/97.1 MB 5.7 MB/s eta 0:00:00


### 3.6. CatBoost

**Гиперпараметр:** `depth` (фиксируем `iterations=100, learning_rate=0.1`)

| depth | test_accuracy | time_sec |
|-------|---------------|----------|
| 3     |               |          |
| 5     |               |          |
| 7     |               |          |
| 10    |               |          |

In [15]:
from catboost import CatBoostClassifier

results_cat = []

# Ваш код здесь (перебор depth, verbose=0)
for depth in [3, 5, 7, 10]:
    start = time.time()
    clf = CatBoostClassifier(depth=depth, verbose=0, iterations=100, learning_rate=0.1, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_cat.append({'depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_cat = pd.DataFrame(results_cat)
print("\nТаблица 7. CatBoost")
print(df_cat.to_markdown(index=False))

depth=3: acc=0.9340, time=25.87s
depth=5: acc=0.9425, time=32.84s
depth=7: acc=0.9442, time=99.77s
depth=10: acc=0.9526, time=755.35s

Таблица 7. CatBoost
|   depth |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|       3 |          0.934  |      25.87 |
|       5 |          0.9425 |      32.84 |
|       7 |          0.9442 |      99.77 |
|      10 |          0.9526 |     755.35 |


## 4. Итоговая таблица лучших результатов

In [16]:
# Соберите лучшие результаты из каждой таблицы вручную или автоматически
# Пример структуры:

summary = pd.DataFrame([
    ('Decision Tree', None, 0.8697, 1.08),  # замените 0 на ваши значения
    ('Random Forest', None, 0.9526, 2.64),
    ('Gradient Boosting', 0.5, 0.9662, 34.65),
    ('XGBoost', 0.5, 0.9492, 31.45),
    ('AdaBoost', 1.5, 0.9255, 6.12),
    ('LightGBM', 15, 0.9662, 4.96),
    ('CatBoost', 10, 0.9526, 755.35),
], columns=['Algorithm', 'Best params', 'Best test acc', 'Time (sec)'])

print("Итоговая таблица лучших результатов")
print(summary.to_markdown(index=False))

Итоговая таблица лучших результатов
| Algorithm         |   Best params |   Best test acc |   Time (sec) |
|:------------------|--------------:|----------------:|-------------:|
| Decision Tree     |         nan   |          0.8697 |         1.08 |
| Random Forest     |         nan   |          0.9526 |         2.64 |
| Gradient Boosting |           0.5 |          0.9662 |        34.65 |
| XGBoost           |           0.5 |          0.9492 |        31.45 |
| AdaBoost          |           1.5 |          0.9255 |         6.12 |
| LightGBM          |          15   |          0.9662 |         4.96 |
| CatBoost          |          10   |          0.9526 |       755.35 |


## 5. Вопросы (ответить текстом в следующей ячейке)

1. Какой алгоритм показал максимальную точность? Какие параметры к этому привели?

2. Какой алгоритм быстрее всего обучался? Во сколько раз он быстрее самого медленного?

3. У каких алгоритмов наблюдалось переобучение? При каких параметрах?

4. Какой алгоритм вы выбрали бы для продакшн-системы с миллионом текстов? Почему?

# Напишите свои ответы здесь

**Ответ 1:**
Максимальную точность показали алгоритмы Gradient Boosting и LightGBM. К такому результату привели следующие параметры: learning_rate=0.5 и num_leaves=15

**Ответ 2:**
Если сравнивать результаты с самой высокой точностью, то быстрее всего обучался алгоритм Decision Tree, он примерно в 755 раз быстрее самого медленного.

**Ответ 3:**
Переобучение наблюдалось у AdaBoost на параметре learning_rate = 2.

**Ответ 4:**
Для продакшн-системы с миллионом текстов я бы выбрала алгоритм LightGBM, потому что он наиболее точный и один из самых быстрых

## Критерии оценки

| Что оценивается | Баллы |
|----------------|-------|
| Заполнены все 7 таблиц (правильно собраны accuracy и время) | 3 |
| Заполнена итоговая таблица лучших результатов | 1 |
| Ответы на 4 вопроса (по 0.5 балла) | 2 |
| Код воспроизводим (random_state=42, порядок ячеек корректен) | 1 |
| **Качество кода** (отсутствие дублирования, осмысленные имена переменных, циклы вместо копипасты, использование .to_markdown()) | 3 |
| **Итого** | **10** |

### Детали по качеству кода (+3 балла):
- **+1** — использование единого шаблона для всех экспериментов (цикл + list.append + pd.DataFrame)
- **+1** — правильное форматирование вывода (округление времени и accuracy до 2-4 знаков)
- **+1** — читаемые имена переменных, комментарии, отсутствие `eval()`, `exec()` и магических чисел

**Штрафы:**
- -1 балл, если код не запускается без ошибок
- -1 балл, если отсутствует `random_state=42` в любой из моделей
- -1 балл, если таблицы выведены криво (не через `to_markdown` или нечитаемый `print`)